File: 2707-extra-characters-in-a-string/2707-extra-characters-in-a-string.py
class Solution:
    def minExtraChar(self, s: str, dictionary: List[str]) -> int:
        cache = {}

        def dfs(text):
            if len(text) <= 0:
                return 0

            if len(text) in cache:
                return cache[len(text)]
            
            minimum = len(text)

            for word in dictionary:
                if text == word:
                    minimum = 0
                    cache[len(text)] = minimum
                    return minimum

                elif text[:len(word)] == word:
                    result = dfs(text[len(word):])
                    minimum = min(minimum, result)

            minimum = min(minimum, 1 + dfs(text[1:]))

            cache[len(text)] = minimum

            return minimum

        return dfs(s)
        

File: 2707-extra-characters-in-a-string/2707-extra-characters-in-a-string.cpp
class Solution {
public:
    int minExtraChar(std::string s, std::vector<std::string> dictionary) {
        std::unordered_map<int, int> cache;

        int result = dfs(s, dictionary, cache);
        return result;
    }

private:
    int dfs(const std::string& text, const std::vector<std::string>& dictionary, std::unordered_map<int, int>& cache) {
        if (text.empty()) {
            return 0;
        }

        if (cache.find(text.length()) != cache.end()) {
            return cache[text.length()];
        }

        int minimum = text.length();

        for (const std::string& word : dictionary) {
            if (text == word) {
                minimum = 0;
                cache[text.length()] = minimum;
                return minimum;
            } else if (text.substr(0, word.length()) == word) {
                int result = dfs(text.substr(word.length()), dictionary, cache);
                minimum = std::min(minimum, result);
            }
        }

        int result = 1 + dfs(text.substr(1), dictionary, cache);
        minimum = std::min(minimum, result);

        cache[text.length()] = minimum;
        return minimum;
    }
};